parustiko/src/lib.rs

```diff
@@ -1,4 +1,65 @@
 mod errors;
 mod protocol;
+mod version_exchange;
 
-pub fn runner() {}
+use crypto::encryption::aes::{GenericArray, AES};
+use crypto::encryption::Encryption;
+use errors::VersionExchangeError;
+use protocol::DecodeRaw;
+use protocol::{BinaryProtocolPacket, Decode};
+use std::io::{Read, Write};
+use std::net::TcpStream;
+use std::time::Duration;
+use version_exchange::{KeyExchange, SshVersion};
+
+// Read bytes from stream until CR and LF ('\r\n') occur
+fn read_header<T: Read>(arr: &mut T) -> Result<Vec<u8>, VersionExchangeError> {
+    let mut header = Vec::with_capacity(51);
+    let mut prev_byte = None;
+
+    loop {
```

Comment:
I don't know if I like this, if we don't find the /r/n we can end up in infinite loop

Reply:
it won't be an infinite loop (if we fail to find \r\n and use the entire stream, the function will end with the EmptyStream error)

Reply:
But theoretically entity can send us a pretty big chunk of data, then we will have to iterate over all of the bytes before we break

Reply:
hmm fair, we could stop iterate after 51 bytes